central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -123,6 +127,115 @@ func (ds *datastoreImpl) addCollectionToGraph(obj *storage.ResourceCollection) e
 	return nil
 }
 
+// graphEdgeUpdates stores added and removed edges from the DAG graph
+type graphEdgeUpdates struct {
+	srcID   string
+	added   []string
+	removed []string
+}
+
+func (ds *datastoreImpl) updateCollectionInGraph(obj *storage.ResourceCollection) (*graphEdgeUpdates, error) {
+	ds.initGraphOnce()
+	ds.graphLock.Lock()
+	defer ds.graphLock.Unlock()
+
+	if obj == nil {
+		return nil, errors.New("passed collection must not be nil")
+	}
+
+	// return object struct
+	ret := &graphEdgeUpdates{
+		obj.GetId(),
+		make([]string, 0),
+		make([]string, 0),
+	}
+
+	// get edges for the current object
+	curEdges, err := ds.graph.GetDescendants(obj.GetId())
+	if err != nil {
+		return nil, errors.Wrap(err, "update attempt on unknown collection")
+	}
+
+	for _, edge := range obj.GetEmbeddedCollections() {
+		_, present := curEdges[edge.GetId()]
+		if !present {
+			ret.added = append(ret.added, edge.GetId())
+		} else {
+			// if we visit an edge, set the value to nil
+			curEdges[edge.GetId()] = nil
+		}
+	}
+
+	// if a value is nil we visited it, if it's not nil we want to remove that edge
+	for key, value := range curEdges {
+		if value != nil {
+			ret.removed = append(ret.removed, key)
+		}
+	}
+
+	// since adding can cause cycles but removing never can, we remove first
+	// we iterate back to front so that if we fail a deletion we can remove the index from our returned list of deletions
+	for idx := len(ret.removed) - 1; idx >= 0; idx-- {
+		deleteErr := ds.graph.DeleteEdge(obj.GetId(), ret.removed[idx])
```

Comment:
If this fails it means that the update operation is unsuccessful, and subsequent operations will have an undesired outcome. Therefore, return err here.